File: projects/02-industrial-equipment/fog/src/main/java/com/fec/industrial/fog/Aggregation.java
package com.fec.industrial.fog;

import java.util.List;

public class Aggregation {

    public record Summary(String sensorType, String siteId, String unit, String windowStart, String windowEnd,
                           int count, double min, double max, double avg, double latest) {}

    public static Summary condenseWindow(String sensorType, String siteId, String unit, List<Reading> readings,
                                  String windowStart, String windowEnd) {
        double min = Double.POSITIVE_INFINITY;
        double max = Double.NEGATIVE_INFINITY;
        double sum = 0;
        for (Reading r : readings) {
            min = Math.min(min, r.value());
            max = Math.max(max, r.value());
            sum += r.value();
        }
        double avg = Math.round((sum / readings.size()) * 1000.0) / 1000.0;
        // Buffer keeps arrival order, so the last element is the most recent sample shown as the live gauge value.
        double latest = readings.get(readings.size() - 1).value();
        return new Summary(sensorType, siteId, unit, windowStart, windowEnd, readings.size(), min, max, avg, latest);
    }
}

File: projects/18-elevator-escalator-fleet-monitoring/fog/aggregation.js
"use strict";

// Pure window-aggregate math over a plain reading array; runs once per group per flush tick.
function rollUpRunWindow(sensorType, siteId, unit, readings, windowStart, windowEnd) {
  const values = readings.map((r) => r.value);
  const sum = values.reduce((total, v) => total + v, 0);
  return {
    sensor_type: sensorType,
    site_id: siteId,
    unit,
    window_start: windowStart,
    window_end: windowEnd,
    count: values.length,
    min: Math.min(...values),
    max: Math.max(...values),
    avg: Math.round((sum / values.length) * 1000) / 1000,
    // latest is the most recent sample, not the window maximum.
    latest: readings[readings.length - 1].value,
  };
}

module.exports = { rollUpRunWindow };
